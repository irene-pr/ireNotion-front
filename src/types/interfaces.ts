export interface IChecklist {
  isChecked: boolean;
  paragraph: string;
}

export interface INote {
  type: string;
  color: string;
  position?: number;
  title?: string;
  paragraph?: string;
  list?: Array<string | IChecklist>;
  id: string;
}

export interface IBoard {
  type: string;
  name: string;
  notes: Array<INote>;
  id: string;
}

export interface IUserContent {
  name: string;
  boards: Array<IBoard>;
}

export interface IUserData {
  userId: string;
  username: string;
}

export interface IState {
  isLoggedIn: boolean;
  userData: IUserData;
  userContent: IUserContent;
}

export interface IUserLoginData {
  username: string;
  password: string;
}

export interface IUserRegisterData {
  name: string;
  username: string;
  password: string;
}
