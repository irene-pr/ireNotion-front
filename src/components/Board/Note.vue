<template>
  <article
    class="note note-paragraph"
    :note="note"
    v-if="note.type === 'paragraph'"
    :class="note.color"
    @dblclick="onClickOpenModal"
  >
    <font-awesome-icon
      icon="pencil-alt"
      class="form__icon-update"
      @click="onClickOpenModal"
    ></font-awesome-icon>
    <font-awesome-icon
      icon="times"
      class="form__icon-close"
      @click="onClickDelete"
    ></font-awesome-icon>
    <h4 v-if="note.title !== ''">{{ note.title }}</h4>
    <p v-if="note.paragraph !== ''">{{ note.paragraph }}</p>
  </article>
  <article
    class="note note-image"
    :note="note"
    v-if="note.type === 'image'"
    :class="note.color"
  >
    <font-awesome-icon
      icon="times"
      class="form__icon-close"
    ></font-awesome-icon>
    <img :src="note.file" alt="" />
  </article>
</template>

<script>
import { defineComponent } from "vue";
import { mapActions } from "vuex";

export default defineComponent({
  name: "Note",
  props: ["note", "idBoard"],
  methods: {
    ...mapActions([
      "deleteNote",
      "updateNote",
      "setUpdateNoteModal",
      "setIdForModal",
    ]),
    onClickDelete() {
      const params = `${this.idBoard}/${this.note.id}`;
      this.deleteNote(params);
    },
    onClickOpenModal() {
      this.setUpdateNoteModal(true);
      this.setIdForModal(this.note.id);
    },
  },
});
</script>

<style lang="scss" scoped>
@import "@/styles/_variables";

.note {
  * {
    margin: 0;
  }

  position: relative;
  width: 100%;
  padding: 25px;
  margin: 10px 0;
  display: flex;
  flex-direction: column;
  font-family: $Lato;
  box-shadow: 0px 5px 7px -4px black;

  @media only screen and (max-width: 480px) {
    width: 90%;
  }

  &-paragraph {
    word-wrap: break-word;
    h4 {
      padding: 0 0 10px 10px;
      font-family: $hand-writing;
      font-weight: bold;
      font-size: 26px;
    }
    p {
      font-family: $hand-writing;
      font-size: 18px;
      font-weight: 100;
    }
  }
  &-image {
    padding: 0;
    img {
      object-fit: contain;
    }
  }
  &:hover {
    .form__icon-close {
      display: block;
    }
    .form__icon-update {
      display: block;
    }
  }
}
.form__icon-close {
  display: none;
  position: absolute;
  right: 10px;
  top: 10px;
  @media only screen and (max-width: 720px) {
    display: block;
  }
}
.form__icon-update {
  display: none;
  position: absolute;
  left: 10px;
  top: 10px;
  font-size: 12px;
  @media only screen and (max-width: 720px) {
    display: block;
  }
}
.orange {
  background-color: $note-orange;
  color: black;
}
.blue {
  background-color: $note-blue;
  color: white;
}
.pink {
  background-color: $note-pink;
  color: white;
}
.yellow {
  background-color: $note-yellow;
  color: black;
}
.green {
  background-color: $note-green;
  color: black;
}
</style>
