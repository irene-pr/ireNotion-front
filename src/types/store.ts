export interface IChecklist {
  isChecked: boolean;
  paragraph: string;
}

export interface INote {
  type?: string;
  color?: string;
  position?: number;
  title?: string;
  paragraph?: string;
  list?: Array<string | IChecklist>;
  id: string;
  userId?: string;
}

export interface IBoard {
  type: string;
  name: string;
  notes: Array<INote>;
  id: string;
}

export interface IUserContent {
  name: string;
  boards: Array<IBoard>;
}

export interface IUserData {
  userId: string;
  username: string;
}

export interface IState {
  isLoggedIn: boolean;
  userContent: IUserContent;
  userData: IUserData;
  isLoading: boolean;
  themeHeaders: string;
  themeSurfaces: string;
  isBoardEditModal: boolean;
  isUpdateNoteModal: boolean;
  isUpdateListNoteModal: boolean;
  isUpdateChecklistNoteModal: boolean;
  idForModal: string;
  listForModal: Array<string>;
}

export interface IThemeState {
  themeHeaders: string;
  themeSurfaces: string;
}

export interface IUserState {
  isLoggedIn: boolean;
  userContent: IUserContent;
  userData: IUserData;
  isLoading: boolean;
}

export interface IModalState {
  isBoardEditModal: boolean;
  isUpdateNoteModal: boolean;
  isUpdateListNoteModal: boolean;
  isUpdateChecklistNoteModal: boolean;
  idForModal: string;
  listForModal: Array<string>;
}
