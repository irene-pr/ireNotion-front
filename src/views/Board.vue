<template>
  <main class="board-page">
    <h1>This is the Board Page!</h1>
  </main>
</template>

<script lang="ts">
import { defineComponent } from "vue";

export default defineComponent({
  name: "Login",
  components: {},
});
</script>

<style lang="scss">
@import "@/styles/_variables";
</style>
