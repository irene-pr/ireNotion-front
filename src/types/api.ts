import { INote } from "./store";

export interface IUserLoginData {
  username: string;
  password: string;
}

export interface IUserRegisterData {
  name: string;
  username: string;
  password: string;
}

export interface IBodyEditNameBoard {
  idBoard: string;
  newName: string;
}
export interface IBodyUpdateNote {
  idNote: string;
  updatedNote: INote;
}
