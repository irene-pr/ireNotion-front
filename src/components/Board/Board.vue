<template>
  <section class="board">
    <div class="board-header">
      <h3 class="board-header__header">{{ board.name }}</h3>
      <div class="board-header__buttons">
        <button class="board-header__button--note">Add Note</button>
        <button class="board-header__button--image">Add Image</button>
      </div>
    </div>

    <draggable class="board__draggable-area" :board="board" @change="log">
      <Note v-for="note in board.notes" :note="note" :key="note.id" />
    </draggable>
  </section>
</template>

<script>
import { defineComponent } from "vue";
import { VueDraggableNext } from "vue-draggable-next";
import Note from "@/components/Board/Note.vue";
import { IBoard } from "@/types/interfaces";

export default defineComponent({
  name: "Board",
  components: {
    draggable: VueDraggableNext,
    Note,
  },
  props: { board: IBoard },
  data() {
    return {
      enabled: true,
      dragging: false,
    };
  },
  methods: {
    log(event) {
      console.log(event);
    },
  },
});
</script>
<style lang="scss" scoped>
.board {
  max-width: 100%;
  width: min-content;
  padding: 15px;
  border: 2px red dotted;
  &__draggable-area {
    width: min-content;
    padding: 15px;
    border: 2px magenta dotted;
  }
  .board-header {
    display: flex;
    justify-content: space-between;
    &__header {
      margin: 0;
      font-family: "Fredericka the Great", cursive;
      font-weight: 100;
      font-size: 28px;
      padding-left: 20px;
    }
    &__buttons {
      display: flex;
      align-content: center;
    }
    &__button {
    }
  }
}
</style>
