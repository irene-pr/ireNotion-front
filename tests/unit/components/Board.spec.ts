import { mount } from "@vue/test-utils";
import { createStore } from "vuex";
import { FontAwesomeIcon } from "@fortawesome/vue-fontawesome";
import { library } from "@fortawesome/fontawesome-svg-core";
import { fas } from "@fortawesome/free-solid-svg-icons";
import { VueDraggableNext } from "vue-draggable-next";
import { IState } from "@/types/interfaces";
import Note from "@/components/Board/Note.vue";
import Board from "@/components/Board/Board.vue";

library.add(fas);

describe("Given a BoardEdit component", () => {
  let mockedState: IState;
  let createParagraphNoteMock = jest.fn();
  let deleteBoardMock = jest.fn();
  let setIdForModalMock = jest.fn();
  let editNameBoardMock = jest.fn();
  let setBoardEditModalMock = jest.fn();
  let onclickAddNoteMock = jest.fn();
  let onClickOpenModalMock = jest.fn();
  let onClickDeleteBoardMock = jest.fn();
  let store = createStore({});
  let options = {
    data() {
      return {
        enabled: true,
        dragging: false,
      };
    },
    components: {
      "font-awesome-icon": FontAwesomeIcon,
      draggable: VueDraggableNext,
      Note,
    },
    stubs: ["FontAwesomeIcon"],
    global: {
      plugins: [store],
    },
    mocks: {
      $store: store,
    },
    props: {
      board: {
        type: "board",
        name: "tablero1",
        notes: [{ type: "paragraph", color: "yellow" }],
        id: "123",
      },
    },
  };
  function getOptions() {
    return {
      data() {
        return {
          enabled: true,
          dragging: false,
        };
      },
      components: {
        "font-awesome-icon": FontAwesomeIcon,
        draggable: VueDraggableNext,
        Note,
      },
      stubs: ["FontAwesomeIcon"],
      global: {
        plugins: [store],
      },
      mocks: {
        $store: store,
      },
      props: {
        board: {
          type: "board",
          name: "tablero1",
          notes: [{ type: "paragraph", color: "yellow" }],
          id: "123",
        },
      },
    };
  }
  function getOptionsWithMethods() {
    return {
      data() {
        return {
          enabled: true,
          dragging: false,
        };
      },
      components: {
        "font-awesome-icon": FontAwesomeIcon,
        draggable: VueDraggableNext,
        Note,
      },
      stubs: ["FontAwesomeIcon"],
      global: {
        plugins: [store],
      },
      methods: {
        onclickAddNote: onclickAddNoteMock,
        onClickOpenModal: onClickOpenModalMock,
        onClickDeleteBoard: onClickDeleteBoardMock,
      },
      mocks: {
        $store: store,
      },
      props: {
        board: {
          type: "board",
          name: "tablero1",
          notes: [{ type: "paragraph", color: "yellow" }],
          id: "123",
        },
      },
    };
  }

  beforeEach(() => {
    mockedState = {
      isLoading: false,
      isLoggedIn: true,
      userContent: {
        name: "Franny",
        boards: [
          {
            type: "board",
            name: "tablero1",
            notes: [{ type: "paragraph", color: "yellow" }],
            id: "123",
          },
          {
            type: "board",
            name: "tablero2",
            notes: [{ type: "paragraph", color: "yellow" }],
            id: "234",
          },
        ],
      },
      userData: {
        userId: "",
        username: "",
      },
      themeHeaders: "night-mode",
      themeSurfaces: "day-mode",
      isBoardEditModal: false,
      isUpdateNoteModal: false,
      idForModal: "",
    };
    createParagraphNoteMock = jest.fn();
    deleteBoardMock = jest.fn();
    setIdForModalMock = jest.fn();
    editNameBoardMock = jest.fn();
    setBoardEditModalMock = jest.fn();
    onclickAddNoteMock = jest.fn();
    onClickOpenModalMock = jest.fn();
    onClickDeleteBoardMock = jest.fn();
    store = createStore({
      state() {
        return mockedState;
      },
      actions: {
        createParagraphNote: createParagraphNoteMock,
        deleteBoard: deleteBoardMock,
        editNameBoard: editNameBoardMock,
        setIdForModal: setIdForModalMock,
        setBoardEditModal: setBoardEditModalMock,
      },
    });
  });
  describe("When it is rendered", () => {
    test("Then it renders", () => {
      options = getOptions();
      mount(Board, options);
    });
    test("Then the it matches the snapshot", () => {
      options = getOptions();
      const wrapper = mount(Board, options);
      expect(wrapper.element).toMatchSnapshot();
    });
    test("Then it renders at least a Note component", () => {
      options = getOptions();
      const wrapper = mount(Board, options);

      wrapper.getComponent(Note);
    });
    test("Then it renders a Header", () => {
      options = getOptions();
      const wrapper = mount(Board, options);

      const header = wrapper.find("h2");

      expect(header.exists()).toBeTruthy();
    });
    test("Then it renders two buttons", () => {
      options = getOptions();
      const wrapper = mount(Board, options);

      const buttons = wrapper.findAll("button");

      expect(buttons).toHaveLength(2);
    });
  });
  describe("When you double click on the Header", () => {
    test("Then the method onClickOpenModal gets called", () => {
      options = getOptionsWithMethods();
      const wrapper = mount(Board, options);

      const header = wrapper.find(".board-header__top");
      header.trigger("dblclick");

      expect(onClickOpenModalMock).toHaveBeenCalled();
    });
  });
  describe("When you click on the pencil icon", () => {
    test("Then the method onClickOpenModal gets called", () => {
      options = getOptionsWithMethods();
      const wrapper = mount(Board, options);

      const pencil = wrapper.find(".form__icon-edit");
      pencil.trigger("click");

      expect(onClickOpenModalMock).toHaveBeenCalled();
    });
    test("Then the action setBoardEditModal gets called", () => {
      options = getOptions();
      const wrapper = mount(Board, options);

      const pencil = wrapper.find(".form__icon-edit");
      pencil.trigger("click");

      expect(setBoardEditModalMock).toHaveBeenCalled();
    });
    test("Then the action setIdForModal gets called", () => {
      options = getOptions();
      const wrapper = mount(Board, options);

      const pencil = wrapper.find(".form__icon-edit");
      pencil.trigger("click");

      expect(setIdForModalMock).toHaveBeenCalled();
    });
  });
  describe("When you click on the cross icon", () => {
    test("Then the method onClickDeleteBoardMock gets called", () => {
      options = getOptionsWithMethods();
      const wrapper = mount(Board, options);

      const cross = wrapper.find(".form__icon-close");
      cross.trigger("click");

      expect(onClickDeleteBoardMock).toHaveBeenCalled();
    });
    test("Then the action deleteBoard gets called", () => {
      options = getOptions();
      const wrapper = mount(Board, options);

      const cross = wrapper.find(".form__icon-close");
      cross.trigger("click");

      expect(deleteBoardMock).toHaveBeenCalled();
    });
  });
  describe("When you click on the add note button", () => {
    test("Then the method onclickAddNote gets called", () => {
      options = getOptionsWithMethods();
      const wrapper = mount(Board, options);

      const addNote = wrapper.find(".board-header__button--note");
      addNote.trigger("click");

      expect(onclickAddNoteMock).toHaveBeenCalled();
    });
    test("Then the actions createParagraphNote gets called", () => {
      options = getOptions();
      const wrapper = mount(Board, options);

      const addNote = wrapper.find(".board-header__button--note");
      addNote.trigger("click");

      expect(createParagraphNoteMock).toHaveBeenCalled();
    });
  });
});
