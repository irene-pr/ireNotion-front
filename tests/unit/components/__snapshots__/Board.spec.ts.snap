// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Board component When it is rendered Then the it matches the snapshot 1`] = `
<section
  class="board"
>
  <div
    class="board-header night-mode"
  >
    <div
      class="board-header__top"
    >
      <h2
        class="board-header__header"
      >
        tablero1
      </h2>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-pencil-alt fa-w-16 form__icon-edit"
        data-icon="pencil-alt"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class=""
          d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
          fill="currentColor"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-times fa-w-11 form__icon-close"
        data-icon="times"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 352 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class=""
          d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      class="board-header__buttons"
    >
      <button
        class="board-header__button board-header__button--note"
        type="button"
      >
         Add Note 
      </button>
      <button
        class="board-header__button board-header__button--image"
        type="button"
      >
         Add Image 
      </button>
    </div>
  </div>
  <div
    board="[object Object]"
    class="board__draggable-area"
  >
    
    
    <article
      class="note note-paragraph yellow"
      note="[object Object]"
    >
      <font-awesome-icon
        class="form__icon-update"
        icon="pencil-alt"
      />
      <font-awesome-icon
        class="form__icon-close"
        icon="times"
      />
      <!--v-if-->
      <!--v-if-->
    </article>
    <!--v-if-->
    
    
  </div>
</section>
`;

exports[`Given a BoardEdit component When it is rendered Then the it matches the snapshot 1`] = `
<section
  class="board"
>
  <div
    class="board-header night-mode"
  >
    <div
      class="board-header__top"
    >
      <h2
        class="board-header__header"
      >
        tablero1
      </h2>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-pencil-alt fa-w-16 form__icon-edit"
        data-icon="pencil-alt"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class=""
          d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
          fill="currentColor"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-times fa-w-11 form__icon-close"
        data-icon="times"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 352 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class=""
          d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      class="board-header__buttons"
    >
      <button
        class="board-header__button board-header__button--note"
        type="button"
      >
         Add Note 
      </button>
      <button
        class="board-header__button board-header__button--image"
        type="button"
      >
         Add Image 
      </button>
    </div>
  </div>
  <div
    board="[object Object]"
    class="board__draggable-area"
  >
    
    
    <article
      class="note note-paragraph yellow"
      note="[object Object]"
    >
      <font-awesome-icon
        class="form__icon-update"
        icon="pencil-alt"
      />
      <font-awesome-icon
        class="form__icon-close"
        icon="times"
      />
      <!--v-if-->
      <!--v-if-->
    </article>
    <!--v-if-->
    
    
  </div>
</section>
`;
